### ISSUE TITLE
`recursion_limit` not respected

### ISSUE THREAD

**Issue by theAkito** [nushell/nushell#13324]
### Describe the bug

Check recursion limit.

```
$env.config.recursion_limit
5000
```

Run recursive procedure.

```
Error: nu::shell::recursion_limit_reached

  × Recursion limit (50) reached
```

### How to reproduce

Set the configuration option, then run your recursive procedure.

### Expected behavior

`recursion_limit` is respected.

### Configuration

| key                | value                                                                |
| ------------------ | -------------------------------------------------------------------- |
| version            | 0.94.2                                                               |
| major              | 0                                                                    |
| minor              | 94                                                                   |
| patch              | 2                                                                    |
| branch             |                                                                      |
| commit_hash        |                                                                      |
| build_os           | linux-x86_64                                                         |
| build_target       | x86_64-unknown-linux-gnu                                             |
| rust_version       | rustc 1.79.0-beta.3 (f5d04caa7 2024-05-03)                           |
| rust_channel       | beta-x86_64-unknown-linux-gnu                                        |
| cargo_version      | cargo 1.79.0-beta.3 (8a7384119 2024-05-03)                           |
| build_rust_channel | release                                                              |
| allocator          | mimalloc                                                             |
| features           | default, sqlite, static-link-openssl, system-clipboard, trash, which |
| installed_plugins  | formats                                                              |

### Additional context

I oriented myself around the following snippet.

https://github.com/merelymyself/nushell/blob/1b291f78b6d9c9f74f791d21c8f73e4afb0b3a7d/crates/nu-utils/src/sample_config/default_config.nu#L241

**Comment by fdncred** [comment]
I'm not sure why this wouldn't be working. Do you have a way to easily reproduce it?
https://github.com/nushell/nushell/blob/399a7c8836488db9fd2601404eb154fddf55e48a/crates/nu-engine/src/eval.rs#L40-L49

**Comment by theAkito** [comment]
> I'm not sure why this wouldn't be working. Do you have a way to easily reproduce it?

```nu
#!/usr/bin/env nu

def recurse [num: int = 0] {
  let numm = $num + 1
  recurse $numm
}

recurse
```

```nu
$ $env.config.recursion_limit
5000
$ $env.config.recursion_limit = 5000
$ nu nushell_13324.nu
Error: nu::shell::recursion_limit_reached

  × Recursion limit (50) reached
   ╭─[/home/user/src/testing/nu/nushell_13324.nu:3:28]
 2 │     
 3 │ ╭─▶ def recurse [num: int = 0] {
 4 │ │     let numm = $num + 1
 5 │ │     recurse $numm
 6 │ ├─▶ }
   · ╰──── This called itself too many times
 7 │     
   ╰────

$ $env.config.recursion_limit
5000
$ 
```

**Comment by theAkito** [comment]
I just upgraded to Nu `0.95.0` and the issue remains the same.

I also `print` the `numm` value now and it gets only to 50, not 5000 iterations.

**Comment by fdncred** [comment]
Thanks. Not sure what's going on.

**Comment by hqsz** [comment]
I also have a hard time reproducing this issue too 😭 

```nu 
> $env.config.recursion_limit
50
> $env.config.recursion_limit = 5000
> $env.config.recursion_limit
5000
~/Workspace/nushell> source nushell_13324.nu

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
[1]    52790 abort      cargo run
```

```
> $env.config.recursion_limit 
50
> $env.config.recursion_limit = 100
> $env.config.recursion_limit
100
> source nushell_13324.nu
Error: nu::shell::recursion_limit_reached

  × Recursion limit (100) reached
   ╭─[.../nushell/nushell_13324.nu:3:28]
 2 │
 3 │ ╭─▶ def recurse [num: int = 0] {
 4 │ │     let numm = $num + 1
 5 │ │     recurse $numm
 6 │ ├─▶ }
   · ╰──── This called itself too many times
 7 │
   ╰────
```

**Comment by fdncred** [comment]
I guess on 5000, it's overflowing before the limit is reached, so it pukes. I have the same behavior on the 100 and, like yours, it says `Recursion limit (100) reached`.

**Comment by devyn** [comment]
>  ```nu
> $ $env.config.recursion_limit
> 5000
> $ $env.config.recursion_limit = 5000
> $ nu nushell_13324.nu
> Error: nu::shell::recursion_limit_reached
> 
>   × Recursion limit (50) reached
>    ╭─[/home/user/src/testing/nu/nushell_13324.nu:3:28]
>  2 │     
>  3 │ ╭─▶ def recurse [num: int = 0] {
>  4 │ │     let numm = $num + 1
>  5 │ │     recurse $numm
>  6 │ ├─▶ }
>    · ╰──── This called itself too many times
>  7 │     
>    ╰────
> 
> $ $env.config.recursion_limit
> 5000
> $ 
> ```

It's not possible to pass `$env.config` to `nu` this way. Environment variables that aren't strings (and maybe a couple of other simple values) don't get propagated to external commands.

You _would_ have to set it within your script, but unfortunately I think we also have another bug where you can't set config in a script and have it reliably affect things that use config, it's only really going to work on the REPL. I'll try to fix that, the config thing's been bugging me for a while anyway.

**Comment by theAkito** [comment]
> It's not possible to pass `$env.config` to `nu` this way. Environment variables that aren't strings (and maybe a couple of other simple values) don't get propagated to external commands.
> 
> You _would_ have to set it within your script, but unfortunately I think we also have another bug where you can't set config in a script and have it reliably affect things that use config, it's only really going to work on the REPL. I'll try to fix that, the config thing's been bugging me for a while anyway.

This was just to "prove" the point of this issue. The configuration is actually set in my `config.nu`, which is sourced on shell opening.

**Comment by theAkito** [comment]
> it's only really going to work on the REPL. I'll try to fix that, the config thing's been bugging me for a while anyway.

Okay, tested it, and at least this works. Well, it crashed the terminal at ~1500 recursions, but that's a different topic... `Thread main is overflown` or something...

**Comment by devyn** [comment]
Yeah, the recursion limit there is actually a guard so that we don't overflow the thread stack. If you overflow the thread stack, it's unrecoverable - nu will just crash. If you want to compile Nushell with a larger stack size, you'll basically have to [pass it at link time like this](https://www.reddit.com/r/rust/comments/872fc4/how_to_increase_the_stack_size/).

**Comment by theAkito** [comment]
> Yeah, the recursion limit there is actually a guard so that we don't overflow the thread stack. If you overflow the thread stack, it's unrecoverable - nu will just crash. If you want to compile Nushell with a larger stack size, you'll basically have to [pass it at link time like this](https://www.reddit.com/r/rust/comments/872fc4/how_to_increase_the_stack_size/).

Sounds all too custom for my use case. I basically just recursively call an API with paginated results, but there's a lot of content to fetch, hence the deep recursion.

I think, in my case, I just need to refactor the code.

However, this issue is, especially based on your comments, still valid in general.

**Comment by fdncred** [comment]
if you look in the nushell repo's .cargo/config.toml file you'll see that we already alter the stack size in windows and there are a couple commented out examples for other operating systems.

**Comment by NotTheDr01ds** [comment]
Also, I would recommend:

```nushell
$env.config.recursion_limit = 500
```

That's generally going to be a safer form than passing in a record, as it will overwrite any other config values. Not a problem in this case, but just a good thing to keep in mind.
